engine/src/main/java/org/pentaho/di/trans/steps/excelinput/staxpoi/StaxPoiWorkbook.java

```diff
@@ -0,0 +1,204 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+/**
+ * Author = Shailesh Ahuja
+ */
+
+package org.pentaho.di.trans.steps.excelinput.staxpoi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+import org.apache.poi.openxml4j.opc.OPCPackage;
+import org.apache.poi.xssf.eventusermodel.XSSFReader;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.logging.KettleLogStore;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.spreadsheet.KSheet;
+import org.pentaho.di.core.spreadsheet.KWorkbook;
+import org.pentaho.di.core.xml.XMLParserFactoryProducer;
+
+/**
+ * Streaming reader for XLSX files.<br>
+ * Does not open XLS.
+ */
+public class StaxPoiWorkbook implements KWorkbook {
+
+  private static final String RELATION_NS_URI = "http://schemas.openxmlformats.org/officeDocument/2006/relationships";
+
+  private LogChannelInterface log;
+
+  private XSSFReader reader;
+
+  // maintain the mapping of the sheet name to its ID
+  private Map<String, String> sheetNameIDMap;
+  // sheet names in order
+  private String[] sheetNames;
+
+  // mapping of the sheet object with its ID/Name
+  private Map<String, StaxPoiSheet> openSheetsMap;
+
+  private OPCPackage opcpkg;
+
+  protected StaxPoiWorkbook() {
+    openSheetsMap = new HashMap<String, StaxPoiSheet>();
+    this.log = KettleLogStore.getLogChannelInterfaceFactory().create( this );
+  }
+
+  public StaxPoiWorkbook( String filename, String encoding ) throws KettleException {
+    this();
+    try {
+      opcpkg = OPCPackage.open( filename );
+      openFile( opcpkg, encoding );
+    } catch ( Exception e ) {
+      throw new KettleException( e );
+    }
+  }
+
+  public StaxPoiWorkbook( InputStream inputStream, String encoding ) throws KettleException {
+    this();
+    try {
+      opcpkg = OPCPackage.open( inputStream );
+      openFile( opcpkg, encoding );
+    } catch ( Exception e ) {
+      throw new KettleException( e );
+    }
+  }
+
+  private void openFile( OPCPackage pkg, String encoding ) throws KettleException {
+    InputStream workbookData = null;
+    XMLStreamReader workbookReader = null;
+    try {
+      reader = new XSSFReader( pkg );
+      sheetNameIDMap = new LinkedHashMap<>();
+      workbookData = reader.getWorkbookData();
+      XMLInputFactory factory = XMLParserFactoryProducer.createSecureXMLInputFactory();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Parsing XML input using a weakly configured XML parser can lead to XML external entity (XXE) injection. To protect against XXE attacks, be sure to securely configure your XML parser.

[Learn more](https://skb.highcastle.a2z.com/implementations/189)